rust/src/mqtt/detect.rs

```diff
@@ -17,421 +17,1528 @@
+        tx,
+        list_id,
+        mqtt_tx_get_publish_message,
+    );
+}
+
+unsafe extern "C" fn mqtt_protocol_version_setup(
+    de: *mut c_void, s: *mut c_void, raw: *const libc::c_char,
+) -> c_int {
+    if DetectSignatureSetAppProto(s, ALPROTO_MQTT) != 0 {
+        return -1;
+    }
+    let ctx = rs_detect_u8_parse(raw) as *mut c_void;
+    if ctx.is_null() {
+        return -1;
+    }
+    if SigMatchAppendSMToList(
+        de,
+        s,
+        G_MQTT_PROTOCOL_VERSION_KW_ID,
+        ctx,
+        G_MQTT_PROTOCOL_VERSION_BUFFER_ID,
+    )
+    .is_null()
+    {
+        mqtt_protocol_version_free(std::ptr::null_mut(), ctx);
+        return -1;
+    }
+    return 0;
+}
+
+unsafe extern "C" fn mqtt_protocol_version_match(
+    _de: *mut c_void, _f: *mut c_void, _flags: u8, state: *mut c_void, _tx: *mut c_void,
+    _sig: *const c_void, ctx: *const c_void,
+) -> c_int {
+    let state = cast_pointer!(state, MQTTState);
+    let ctx = cast_pointer!(ctx, DetectUintData<u8>);
+    if detect_match_uint(ctx, state.protocol_version) {
+        return 1;
+    }
+    return 0;
+}
+
+unsafe extern "C" fn mqtt_protocol_version_free(_de: *mut c_void, ctx: *mut c_void) {
+    let ctx = cast_pointer!(ctx, DetectUintData<u8>);
+    rs_detect_u8_free(ctx);
+}
+
+// maybe to factor with websocket.flags
+struct MqttParsedFlagItem {
+    neg: bool,
+    value: u8,
+}
+
+fn parse_flag_list_item(s: &str) -> IResult<&str, MqttParsedFlagItem> {
+    let (s, _) = opt(is_a(" "))(s)?;
+    let (s, neg) = opt(tag("!"))(s)?;
+    let neg = neg.is_some();
+    let (s, value) = alt((value(0x8, tag("dup")), value(0x1, tag("retain"))))(s)?;
+    let (s, _) = opt(is_a(" ,"))(s)?;
+    Ok((s, MqttParsedFlagItem { neg, value }))
+}
+
+fn parse_flag_list(s: &str) -> IResult<&str, Vec<MqttParsedFlagItem>> {
+    return many1(parse_flag_list_item)(s);
+}
+
+fn parse_flags(s: &str) -> Option<DetectUintData<u8>> {
+    // try first numerical value
+    if let Ok((_, ctx)) = detect_parse_uint::<u8>(s) {
+        return Some(ctx);
+    }
+    // otherwise, try strings for bitmask
+    if let Ok((rem, l)) = parse_flag_list(s) {
+        if !rem.is_empty() {
+            SCLogWarning!("junk at the end of mqtt.flags");
+            return None;
+        }
+        let mut arg1 = 0;
+        let mut arg2 = 0;
+        for elem in l.iter() {
+            if elem.value & arg1 != 0 {
+                SCLogWarning!("Repeated bitflag for mqtt.flags");
+                return None;
             }
-            MQTTOperation::DISCONNECT(ref v) => {
-                if let Some(rcode) = v.reason_code {
-                    *result = rcode;
-                    return 1;
-                }
+            arg1 |= elem.value;
+            if !elem.neg {
+                arg2 |= elem.value;
             }
-            _ => return 0,
         }
+        let ctx = DetectUintData::<u8> {
+            arg1,
+            arg2,
+            mode: DetectUintMode::DetectUintModeBitmask,
```

Comment:
Nice. This is surely powerful, but really needs documentation or examples. Took me a bit to get my head around it!

Reply:
Oh I was referring to the Rust API (`DetectUintMode` and what tooling the Rust helpers provide). I think for common cases like negatable flags it would be helpful to see an example so the wheel will not be reinvented again ;)